src/nvim/ex_getln.c

```diff
@@ -2161,6 +2175,14 @@ static void draw_cmdline(int start, int len)
 {
   int i;
 
+  if (cmdline_external) {
+    Array args = ARRAY_DICT_INIT;
+    ADD(args, STRING_OBJ(cstr_to_string((char *)(ccline.cmdbuff))));
+    ADD(args, INTEGER_OBJ(ccline.cmdpos));
```

Comment:
At least to start with I think we can assume one single word (sequence of codepoints being shaped together) of arabic text will have just one color. So at the end of the  `if arabic_char()` block we could check if the byte index is equal to or greater than the end of the current highlight chunk and then "flush" the arshape_buf so far with `msg_outtrans_len_attr`